### PR TITLE
feat(frontend/shared): add chat functionality tm-112

### PR DESCRIPTION
**Pull request by Ann-Maksay**
According to the ticket it's possible:
- Create a chat.
- View all existing chats and messages in them.
- Send a message.

Demo:
create-chat2
answer-chat
show-chat

UI elements:
chats-link
empty-chat
chat
message-pending

### CODE REVIEW

File: apps/frontend/src/assets/img/chat-img.png
Comment (what1s1ove) [inline apps/frontend/src/assets/img/chat-img.png:1]: use `svg` instead

File: apps/frontend/src/libs/constants/date-format.constants.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const DATE_FORMAT_VALUES = {
```
Comment (what1s1ove) [inline apps/frontend/src/libs/constants/date-format.constants.ts:1]: B29

File: apps/frontend/src/libs/helpers/create-query-link/create-query-link.helper.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { type AppRoute } from "~/libs/enums/enums.js";
+import { type ValueOf } from "~/libs/types/types.js";
+
+const createQueryLink = (
```
Comment (what1s1ove) [inline apps/frontend/src/libs/helpers/create-query-link/create-query-link.helper.ts:4]: you need to use the URL/URLSearchParams constructor instead and i am not sure if you really need to have this helper

File: apps/frontend/src/modules/app/app.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export { actions, reducer } from "./slices/app.js";
```
Comment (what1s1ove) [inline apps/frontend/src/modules/app/app.ts:1]: we will not have such module in the APP move it to `src/libs/slices/app`

File: apps/frontend/src/libs/helpers/format-date/format-date.helper.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { DATE_FORMAT_VALUES } from "~/libs/constants/constants.js";
+
+const formatDate = (inputDate: string): string => {
```
Comment (what1s1ove) [inline apps/frontend/src/libs/helpers/format-date/format-date.helper.ts:3]: let's install `date-fns` package in the shared module and we will use it instead

File: apps/frontend/src/modules/chats/slices/chats.slice.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import { type ValueOf } from "~/libs/types/types.js";
+import { type UserAuthResponseDto } from "~/modules/users/users.js";
+
+import { type ChatGetAllItemResponseDto } from "../libs/types/types.js";
+import { createChat, getAllChats, getChat } from "./actions.js";
+
+type State = {
+	chats: ChatGetAllItemResponseDto[];
+	dataStatus: ValueOf<typeof DataStatus>;
+	interlocutor: UserAuthResponseDto | null;
```
Comment (what1s1ove) [inline apps/frontend/src/modules/chats/slices/chats.slice.ts:13]: can we use user from `auth` slice instead?

File: apps/frontend/src/pages/chats/libs/components/chat-sidebar/chat-sidebar.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import styles from "./styles.module.css";
+
+type Properties = {
+	chats: ChatGetAllItemResponseDto[];
+};
+
+const ChatSidebar: React.FC<Properties> = ({ chats }: Properties) => {
+	return (
+		<ul className={styles["chats"]}>
+			{Boolean(chats) && chats.length > ARRAY_EMPTY_LENGTH ? (
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/components/chat-sidebar/chat-sidebar.tsx:16]:
```suggestion
const hasMessages = Boolean(chats.length)
			{hasMessages && (
```

File: apps/frontend/src/pages/chats/libs/components/chat-sidebar/chat-sidebar.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+				chats.map((item) => {
+					return (
+						<li key={item.id}>
+							<ChatLink chat={item} />
+						</li>
+					);
+				})
+			) : (
+				<p className={styles["empty-chat-list"]}>
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/components/chat-sidebar/chat-sidebar.tsx:25]: only `li` elements can be inside the `ul` element

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/chats/libs/components/chat-sidebar/chat-sidebar.tsx (hunk 6 above)
Comment (Ann-Maksay) [reply apps/frontend/src/pages/chats/libs/components/chat-sidebar/chat-sidebar.tsx:16]: Linter doesn't allow to do it without 0
image

File: apps/frontend/src/modules/chats/libs/types/types.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+export {
+	type ChatCreateRequestDto,
+	type ChatGetAllItemResponseDto,
+	type ChatSingleItemResponseDto,
```
[the remark this answers is not in the document]
Comment (ShadowOfTheSpace) [reply apps/frontend/src/modules/chats/libs/types/types.ts:4]: I changed it to `ChatItemResponseDto` in my PR too

File: apps/frontend/src/pages/chats/libs/components/chat/styles.module.css [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	width: 5px;
+}
+
+.chat-container::-webkit-scrollbar-thumb {
+	background-color: var(--color-text-100);
+	border-radius: 50px;
+}
+
+.chat-container::-webkit-scrollbar-track {
+	background-color: #ffffff;
```
Comment (YourTeaChan) [inline apps/frontend/src/pages/chats/libs/components/chat/styles.module.css:43]:
```suggestion
	background-color: var(--main-white);
```

File: apps/frontend/src/pages/chats/libs/components/chat-sidebar/styles.module.css [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	width: 5px;
+}
+
+.chats::-webkit-scrollbar-thumb {
+	background-color: var(--color-text-100);
+	border-radius: 50px;
+}
+
+.chats::-webkit-scrollbar-track {
+	background-color: #ffffff;
```
Comment (YourTeaChan) [inline apps/frontend/src/pages/chats/libs/components/chat-sidebar/styles.module.css:20]:
```suggestion
	background-color: var(--main-white);
```

File: apps/frontend/src/pages/chats/libs/components/chat-form/styles.module.css [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+	width: 5px;
+}
+
+.input::-webkit-scrollbar-thumb {
+	background-color: var(--color-text-100);
+	border-radius: 50px;
+}
+
+.input::-webkit-scrollbar-track {
+	background-color: #ffffff;
```
Comment (YourTeaChan) [inline apps/frontend/src/pages/chats/libs/components/chat-form/styles.module.css:24]:
```suggestion
	background-color: var(--main-white);
```

File: apps/frontend/src/pages/chats/libs/components/chat-message/chat-message.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+const ChatMessage: React.FC<Properties> = ({
+	isCurrentUserSender,
+	message,
+}: Properties) => {
+	const { createdAt, senderUser, text } = message;
+
+	const contsinerClassNmaes = isCurrentUserSender
+		? getValidClassNames(styles["container"], styles["right"])
+		: getValidClassNames(styles["container"], styles["left"]);
```
Comment (ShadowOfTheSpace) [inline apps/frontend/src/pages/chats/libs/components/chat-message/chat-message.tsx:19]:
```suggestion
	const containerClassNames = getValidClassNames(styles["container"], styles[isCurrentUserSender ? "right" : "left"])
```

**Review by ShadowOfTheSpace — APPROVED** [review]

**Review by what1s1ove — CHANGES_REQUESTED** [review]
Please repoint your branch to `main` and resolve conflicts

**Review by YourTeaChan — APPROVED** [review]

File: apps/frontend/src/assets/icons/message.svg
Comment (Ann-Maksay) [inline apps/frontend/src/assets/icons/message.svg:1]: hover

File: apps/frontend/src/pages/chats/libs/components/empty-chat/empty-chat.tsx [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import chatImage from "~/assets/img/chat-character.svg";
+import { Image, Link } from "~/libs/components/components.js";
+import { AppRoute } from "~/libs/enums/app-route.enum.js";
```
Comment (GEOFARL) [inline apps/frontend/src/pages/chats/libs/components/empty-chat/empty-chat.tsx:3]:
```suggestion
import { AppRoute } from "~/libs/enums/enums.js";
```

File: apps/frontend/src/modules/chats/slices/chats.slice.ts (hunk 5 above)
Comment (what1s1ove) [reply apps/frontend/src/modules/chats/slices/chats.slice.ts:13]: got it, thank you!

File: packages/shared/src/libs/enums/date-value.enum.ts [hunk 14]
```diff
@@ -0,0 +1,5 @@
+const DateValue = {
+	HOURS_IN_DAY: 24,
+};
```
Comment (what1s1ove) [inline packages/shared/src/libs/enums/date-value.enum.ts:3]:
```suggestion
} as const;
```

File: packages/shared/src/libs/helpers/format-date/format-date.helper.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+import { differenceInHours, format } from "date-fns";
+
+import { DateValue } from "../../enums/enums.js";
+
+const formatDate = (inputDateString: string): string => {
```
Comment (what1s1ove) [inline packages/shared/src/libs/helpers/format-date/format-date.helper.ts:5]:
```suggestion
const formatDate = (inputDateString: string, type: valuesof<FormatDateType> ): string => {
```
add an enum for it and use it to pass to `format` function

Comment (what1s1ove) [reply packages/shared/src/libs/helpers/format-date/format-date.helper.ts:5]: I am not even sure that you will need a `DateValue` then

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by GEOFARL — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**